R/mobr_boxplots.R

```diff
@@ -33,6 +33,130 @@ calc_PIE = function(x) {
+
+   if (length(abund_vec) > 0){
+      out_vec["N"] <- sum(abund_vec)
+      out_vec["S"] <- length(abund_vec)
+      if (!is.na(n_rare))
+         out_vec["S_rare"] <- rarefaction(abund_vec, method = "indiv", effort = n_rare)
+      out_vec["PIE"] <- diversity(abund_vec, index = "simpson")
+      out_vec["ENS_PIE"] <- diversity(abund_vec, index = "invsimpson")
+      
+      S_asymp <- try(vegan::estimateR(abund_vec))
+      if (class(S_asymp) == "try_error"){
+         warning("The Chao richness estimator cannot be calculated for all groups.")
+      } else {
+         out_vec["S_asymp"] = S_asymp["S.chao1"]
+      }
+   }
+   
+   return(out_vec)
+}
+
+# generate a single bootstrap sample of group-level biodiversity indices
+boot_sample_group <- function(comm_dat, n_rare)
+{
+   n_sample <- nrow(comm_dat)
+   sample_rows <- sample(1:n_sample, size = n_sample, replace = T)
+   abund <- colSums(comm_dat[sample_rows, ])
+   out <- calc_biodiv_single_group(abund, n_rare = n_rare)
+   
+   return(out)
+}
+
+# replicated bootstrap samples of group-level indices
+repl_boot_samples <- function(comm_dat, n_rare, n_perm = 100)
+{
+   repl_biodiv <- replicate(n_perm, boot_sample_group(comm_dat, n_rare = n_rare))
+   return(repl_biodiv)
+}
+
+# get confidence interval from bootstrap samples
+get_CI <- function(x, level = 0.95)
+{
+   alpha <- 1 - level
+   p <- c(alpha/2, 1 - alpha/2)
+   CI_tab <- apply(x, MARGIN = 1, quantile, probs = p)
+   CI_vec <- c(CI_tab[1,], CI_tab[2,])
+   names(CI_vec) <- paste(rep(colnames(CI_tab), 2),
+                          rep(c("low", "up"), each = ncol(CI_tab)), sep = "_")
+   return(CI_vec)
+}
+
+
+
+# # Calculate group-level biodiversity indices
+# # http://www.uni-kiel.de/psychologie/rexrepos/posts/resamplingBootStrat.html
+# 
+# calc_group_indices <- function(comm_dat, idx, n_min = 10){
+#    
+#    group = comm_dat[idx,1]
+#    abund_dat = comm_dat[idx,-1]
+#    
+#    abund_group = aggregate(abund_dat, by = list(group), FUN = "sum")[ ,-1]
+#    
+#    # Group-based statistics
+#    N_group = rowSums(abund_group)      
+#    S_group = rowSums(abund_group > 0) 
+#    
+#    N_min_group = max(n_min, min(N_group))
+#    groups_low_n = N_group < n_min
+#    
+#    S_rare_group = rep(NA, length(N_group))
+#    S_rare_group[!groups_low_n] = apply(abund_group[!groups_low_n], MARGIN = 1,
+#                                        FUN = rarefaction, method = "indiv",
+#                                        effort = N_min_group)
+# 
+#    # Probability of Interspecific Encounter
+#    groups_n0 = N_group == 0
+#    
+#    
+#    PIE_group = diversity(abund_group, index = "simpson")
+#    ENS_PIE_group = diversity(abund_group, index = "invsimpson")
+#    PIE_group[groups_n0] = NA
+#    ENS_PIE_group[groups_n0] = NA
+#    
+#    # bias corrected Chao estimator
+#    S_asymp_group = estimateR(abund_group)["S.chao1",]
+#    
+#    # bias corrected Chao estimator
+#    S_asymp_group = try(vegan::estimateR(abund_group))
```

Comment:
Can you add comments for why `S_asump_group` is defined twice. I suppose we should just keep one of these lines...?